### PR TITLE
GC IPL Hack for Configurable MEM1 Size.

### PR DESCRIPTION
**Pull request by Minty-Meeo**
A follow up to my prior Pull Request (Configurable MEM1 and MEM2 sizes at runtime via Dolphin.ini #8722).  This is a hack to assembly edit retail IPLs to support different MEM1 sizes before they are copied into Emulator Memory.  As with any other Emulated Memory Size Overrides, this change does nothing if the MAIN_RAM_OVERRIDE_ENABLE config setting is false.
I also updated a comment in Memmap.cpp to reflect new behavior from this commit and my prior one as well.

### CODE REVIEW

File: Source/Core/Core/Boot/Boot.cpp [hunk 1]
```diff
@@ -351,6 +351,44 @@ bool CBoot::Load_BS2(const std::string& boot_rom_filename)
   // Run the descrambler over the encrypted section containing BS1/BS2
   ExpansionInterface::CEXIIPL::Descrambler((u8*)data.data() + 0x100, 0x1AFE00);
 
+  // Perform a surgical modification on known IPLs for Emulated Memory Size Override.
+  if (Config::Get(Config::MAIN_RAM_OVERRIDE_ENABLE))
```
Comment (lioncash) [inline Source/Core/Core/Boot/Boot.cpp:355]: This block should go in its own function so all of the memory override stuff is self contained.

File: Source/Core/Core/Boot/Boot.cpp [hunk 2]
```diff
@@ -351,6 +351,44 @@ bool CBoot::Load_BS2(const std::string& boot_rom_filename)
...
+    u32* ori_that_ruins_things;
+
+    switch (ipl_hash)
+    {
+    case NTSC_v1_0:
+    case NTSC_v1_1:
+    case PAL_v1_0:
+    case MPAL_v1_1:
+      lis_simm_field = (s16*)(data.data() + 0xCB6);         // 0x820 + 0x494 + 2
+      ori_that_ruins_things = (u32*)(data.data() + 0x5FC);  // 0x100 + 0x4fc
```
Comment (lioncash) [inline Source/Core/Core/Boot/Boot.cpp:374]: Dereferencing any of these will result in undefined behavior, as this violates C++ strict aliasing rules. If you want to make these modifications, then you need to make use of `std::memcpy` to do it in a well-defined manner e.g.

```cpp
const u32 nop = Common::swap16(0x60000000);
std::memcpy(data.data() + 0x5FC, &nop, sizeof(nop));
std::memcpy(data.data() + 0xCB6, &mem1_size, sizeof(mem1_size));
```

File: Source/Core/Core/Boot/Boot.h [hunk 3]
```diff
@@ -162,3 +170,5 @@ void UpdateStateFlags(std::function<void(StateFlags*)> update_function);
 /// Normally, this is automatically done by ES when the System Menu is installed,
 /// but we cannot rely on this because we don't require any system titles to be installed.
 void CreateSystemMenuTitleDirs();
+
+void RAMOverrideForIPLMemoryValues(std::string& data, const u32 ipl_hash);
```
Comment (lioncash) [inline Source/Core/Core/Boot/Boot.h:174]: This doesn't need to be exposed. This function can be an internal static function. It just needs to be defined above its usage spot.

File: Source/Core/Core/Boot/Boot.h [hunk 4]
```diff
@@ -32,6 +32,14 @@ struct RegionSetting
   std::string code;
 };
 
+// CRC32 hashes of the IPL file, obtained from Redump
+constexpr u32 NTSC_v1_0 = 0x6DAC1F2A;
```
Comment (lioncash) [inline Source/Core/Core/Boot/Boot.h:36]: Why were these moved out here? These can be kept within the cpp file.

File: Source/Core/Core/Boot/Boot.cpp [hunk 5]
```diff
@@ -593,3 +588,38 @@ void CreateSystemMenuTitleDirs()
   const auto es = IOS::HLE::GetIOS()->GetES();
   es->CreateTitleDirectories(Titles::SYSTEM_MENU, IOS::SYSMENU_GID);
 }
+
+void RAMOverrideForIPLMemoryValues(std::string& data, const u32 ipl_hash)
+{
+  if (Config::Get(Config::MAIN_RAM_OVERRIDE_ENABLE))
```
Comment (lioncash) [inline Source/Core/Core/Boot/Boot.cpp:594]: Now that this is localized within its own function, this can be inverted like so to allow unindenting all of the code within it.

```cpp
if (!Config::Get(Config::MAIN_RAM_OVERRIDE_ENABLE))
  return;
```

File: Source/Core/Core/Boot/Boot.h (hunk 4 above)
Comment (Minty-Meeo) [reply Source/Core/Core/Boot/Boot.h:36]: Good point

File: Source/Core/Core/Boot/Boot.cpp [hunk 6]
```diff
@@ -300,19 +308,46 @@ bool CBoot::LoadMapFromFilename()
   return false;
 }
 
+void RAMOverrideForIPLMemoryValues(std::string& data, const u32 ipl_hash)
```
Comment (lioncash) [inline Source/Core/Core/Boot/Boot.cpp:311]:
```suggestion
static void RAMOverrideForIPLMemoryValues(std::string& data, const u32 ipl_hash)
```

Otherwise this results in a -Wmissing-declaration warning.

File: Source/Core/Core/Boot/Boot.cpp [hunk 7]
```diff
@@ -300,19 +308,46 @@ bool CBoot::LoadMapFromFilename()
...
+
+  // Pre-calculate the new values.
+  const s16 mem1_s_size = Common::swap16(Config::Get(Config::MAIN_MEM1_SIZE) >> 16);
+  const u32 nop = Common::swap32(0x60000000);
+
+  // [lis	r0, 0x0180] <-- This instruction is what sets the Simulated MEM1 Size in retail IPLs.
+  // The following code replaces the SIMM field and nothing else.
+  // [ori	r9, r9, 0x0003] <-- This instruction putting the value "3" into r9 causes a stw later
+  // down the line to revert the instruction we're trying to change.  I don't understand why, but
+  // I do know that NOPing this away doesn't break anything!
```
Comment (Pokechu22) [inline Source/Core/Core/Boot/Boot.cpp:324]: It appears that `r9` is then written to [EXI0CR](http://hitmen.c02.at/files/yagcd/yagcd/chap5.html#sec5) (`0xCC00680C`), where 3 is start (1) DMA (2) read.  Setting this to 0 disables it (and since the code then polls EXI0CR for bit 0 to be cleared, it immediately finishes).

But perhaps it would be better to modify the data it's reading from EXI instead of patching it out?  I'm not sure where it'd be getting it from even (maybe it's something in `EXI_DeviceIPL`, but I don't see anything that explicitly implements it).

Comment (Minty-Meeo) [reply Source/Core/Core/Boot/Boot.cpp:324]: Finally some closure on what the heck was going on.  I was so confused watching as a stw instruction writing to some bizarre address was causing instructions to revert of all things.

Comment (Pokechu22) [reply Source/Core/Core/Boot/Boot.cpp:324]: ... in fact, the whole `Memory::CopyToEmu(0x01300000, data.data() + 0x820, 0x1AFE00);` line is unneeded, since BS1 does that itself ([see YAGCD](http://hitmen.c02.at/files/yagcd/yagcd/chap18.html#sec18.2), though when they say "clearing bit 17 of 0xcc006800" it looks like it should be bit 13 (`ROMDIS`)).

But I don't think you need to address that in this PR; just change the comment here to something like

> // [ori	r9, r9, 0x0003] <-- The value 3 in r9 later starts an EXI DMA read to copy BS2, which clobbers the patched version.  Leaving it at 0 makes it think the EXI transfer finished immediately, skipping it.

File: Source/Core/Core/Boot/Boot.cpp [hunk 8]
```diff
@@ -300,19 +308,46 @@ bool CBoot::LoadMapFromFilename()
...
+static void RAMOverrideForIPLMemoryValues(std::string& data, const u32 ipl_hash)
+{
+  if (!Config::Get(Config::MAIN_RAM_OVERRIDE_ENABLE))
+    return;
+
+  // Pre-calculate the new values.
+  const s16 mem1_s_size = Common::swap16(Config::Get(Config::MAIN_MEM1_SIZE) >> 16);
+  const u32 nop = Common::swap32(0x60000000);
+
+  // [lis	r0, 0x0180] <-- This instruction is what sets the Simulated MEM1 Size in retail IPLs.
```
Comment (Pokechu22) [inline Source/Core/Core/Boot/Boot.cpp:320]: As far as I can tell, it's writing to `80000028`, which [wiibrew calls](https://wiibrew.org/wiki/Memory_Map) is "Memory Size (Physical)", not Simulated MEM1 size.  Simulated Memory Size is 800000F0 instead (which doesn't appear to be written by the IPL).

Comment (Minty-Meeo) [reply Source/Core/Core/Boot/Boot.cpp:320]: My mistake, I just assumed it was simulated.

**Comment by Minty-Meeo** [comment]
lol I thought it got merged already I'm a dummy

**Comment by RenaKunisaki** [comment]
Just curious why this doesn't affect the simulated size? And MEM2 size seems to do nothing, couldn't it change ARAM size?

**Comment by Minty-Meeo** [comment]
> 
> 
> Just curious why this doesn't affect the simulated size? And MEM2 size seems to do nothing, couldn't it change ARAM size?

Changing MEM2 for GameCube does nothing since it doesn't have MEM2.  Changing ARAM should be possible but I didn't target that with my PR that added configurable RAM sizes since I wasn't aware that could be changed at the time.
I've been away from Dolphin devving for a bit since I'm busy with a romhack, but I will come back to this eventually.

**Comment by Minty-Meeo** [comment]
It's time to bring this PR back to life.

**Comment by Minty-Meeo** [comment]
I am exploring potentially emulating the IPL's presence in physical memory accurately and I might roll back re-ordering HW::Init() in the way I did.

File: Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp [hunk 9]
```diff
@@ -98,31 +100,90 @@ void CEXIIPL::Descrambler(u8* data, u32 size)
...
+      break;
+    case CRC32_NTSC_v1_2:
+    case CRC32_PAL_v1_2:
+      std::memcpy(data + 0x820 + 0x490 + 2, &simm_p_mem1_size, sizeof(simm_p_mem1_size));
+      std::memcpy(data + 0x100 + 0x31C + 2, &simm_p_mem1_size, sizeof(simm_p_mem1_size));
+      std::memcpy(data + 0x100 + 0x36C + 2, &simm_p_mem1_size, sizeof(simm_p_mem1_size));
+      break;
+    default:
+      PanicAlertFmtT(
+          "Emulated Memory Size Override is not supported for unrecognized IPL dumps.\n");
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp:128]: Does this intentionally add the newline at the end? There's nothing behind it, and the dialog ends; so it is either a left-over or you forgot to add something _else_ there.

File: Source/Core/Core/Boot/Boot.cpp [hunk 10]
```diff
@@ -594,18 +520,46 @@ bool CBoot::BootUp(std::unique_ptr<BootParameters> boot)
...
+          ExpansionInterface::GetChannel(0)->GetDevice(2));
+      DEBUG_ASSERT(exi_ipl->m_device_type == ExpansionInterface::EXIDeviceType::MaskROM);
+
+      if (!exi_ipl->IsROMLoaded())
       {
         if (ipl.disc)
           PanicAlertFmtT("Cannot start the game, because the GC IPL could not be found.");
         else
-          PanicAlertFmtT("Cannot find the GC IPL.");
+          PanicAlertFmtT("Could not find the GC IPL.");
```
Comment (BhaaLseN) [inline Source/Core/Core/Boot/Boot.cpp:535]: Seems a bit odd that you only changed one branch of the check, and with a very minor change ("Cannot" vs. "Could not").

Any practical reason why past tense is more appropriate here than the previous present tense?

File: Source/Core/Core/Boot/Boot.cpp [hunk 11]
```diff
@@ -594,18 +520,46 @@ bool CBoot::BootUp(std::unique_ptr<BootParameters> boot)
...
         return false;
       }
 
-      if (!Load_BS2(ipl.path))
-        return false;
+      // MSR.IP = 1;
```
Comment (BhaaLseN) [inline Source/Core/Core/Boot/Boot.cpp:539]: Should this be further down with the other `MSR` initializations, and uncommented?

File: Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp (hunk 9 above)
Comment (Minty-Meeo) [reply Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp:128]: Oh right.  This is a shortened copy of a message I had in a previous version of this PR.  There used to be a second line of dialogue, yes.  I'll fix that.

File: Source/Core/Core/Boot/Boot.cpp (hunk 10 above)
Comment (Minty-Meeo) [reply Source/Core/Core/Boot/Boot.cpp:535]: I found it appropriate to use past tense here, since loading the IPL now takes place earlier, in ExpansionInterface::Init().

File: Source/Core/Core/Boot/Boot.cpp (hunk 11 above)
Comment (Minty-Meeo) [reply Source/Core/Core/Boot/Boot.cpp:539]: According to yagcd, MSR.IP is set to 1 when the GameCube is powered on.  It's not part of the HLE hack that is being used, but I thought I wanted to document it somehow.  I think I'll remove this, now, since I want to try putting the IPL at physical memory 0xFFF00000 for a future PR.

**Comment by Adamillo** [comment]
I don't quite understand this as a casual Dolphin user that likes to do dumb experiments. What is this feature about?

**Comment by JosJuice** [comment]
Adding more memory does not improve the performance.

**Comment by Minty-Meeo** [comment]
> I don't quite understand this as a casual Dolphin user that likes to do dumb experiments. What is this feature about?

Emulated Memory Override was made possible because CBoot::EmulatedBS2_GC can set the physical MEM1 size in Dolphin-OS globals to arbitrary values. It is incompatible with retail IPLs since they hardcode the physical MEM1 size to be 24MiB.  This PR tackles the niche use case of combining Emulated Memory Override with running the real IPL.  By assembly editing the SIMM fields containing the value 24MiB, now retail IPLs can join the fun.

Emulated Memory Override is only really useful for a leaked Metroid Prime 3 GCN build that ran on dev hardware (they had 48MiB), maybe one Fire Emblem game that has random crashes in Dolphin, and any romhacks willing to give up console compatibility for additional RAM.

**Comment by Minty-Meeo** [comment]
I was reminded that the TDEV IPL exists and is configured for 48MiB of physical MEM1.  This got me thinking that maybe this feature should be a seperate toggle from Emulated Memory Override.  Thoughts?  Where in the GUI should this even go?  Also, how can I get verified hashes for these obscure IPLs?

**Comment by Pokechu22** [comment]
>  Also, how can I get verified hashes for this obscure IPL and others?

I believe Redump's BIOS datfile from [here](http://redump.org/downloads/) contains the hashes for those now (though it didn't in the past)

**Comment by Minty-Meeo** [comment]
I have found something interesting in the NPDP GDEV IPL.  It checks bit 3 of the DVD Interface Configuration Register (0xCC006024) which YAGCD says is unused.  If it is 0, the p mem1 size is set to 48MiB.  If it is 1, the p mem1 size is set to 24MiB.  Dolphin emulates it such that the 48MiB path is chosen.  I'm not sure how I should patch this IPL?

**Comment by Minty-Meeo** [comment]
While everything in this PR is functional, I know it edges on being outside of Dolphin Emulator's scope.  If it gets accepted, awesome.  If it gets rejected, I'll bring the practical changes to a future PR.  I don't think there is anything more to change.

**Comment by Minty-Meeo** [comment]
I have finished writing a command line tool in Python that does everything this Pull Request can.  Unless there are any objections, I think I will close this pull request tonight and bring the practical improvements to a future PR like I said before.  While this PR is neat, I see now that it is textbook feature-creep.  Thank you all for your help and guidance.
